csrc/device_lower/analysis/trivial_broadcast.cpp

```diff
@@ -95,6 +92,14 @@ void ConcretizedBroadcastDomains::dispatch(Expr* expr) {
         const bool is_concretized =
             !c_id->isBroadcast() && !c_id->isReduction();
         auto it = broadcast_origin_map_.find(p_id);
+        if (it == broadcast_origin_map_.end()) {
+          // The consumer might be broadcast even though the producer is not, if
+          // this is a pad to size 1. In such case, we set the consumer as the
+          // origin of the broadcast.
+          broadcast_origin_map_.emplace(
```

Comment:
Why `c_id` is added here? It may not be a broadcast domain, right?

I think we would need to look at each tensor and see if it has an rfactor broadcast domain to look for a new broadcast.

Reply:
You're right. I think this might have only worked since the only time the condition was being hit was for pads to broadcast. In that case there is only a single producer so this is fine, although we should probably still check `c_id->isBroadcast()`.

I think currently the only time we'll wind up with a `Broadcast` ID without a `BroadcastOp` is from a `PadOp`. Reshapes explicitly broadcast if a target size is 1, and `cat` appears with upstream `PadOp`s in the IR which would handle the broadcasting. The current approach in this PR is kind of a landmine; if we introduce a new op that can produce a broadcast we can hit this cryptic error again. I'm don't know how to handle it more elegantly yet.

Reply:
Doesn't `SliceOp` also possibly produce broadcast domains?

> if we introduce a new op that can produce a broadcast we can hit this cryptic error again. I'm don't know how to handle it more elegantly yet.

I wonder why this check doesn't get triggered when a broadcast is produced by a pad op, which is then resolved:

```
    NVF_ERROR(
            it != broadcast_origin_map_.end(),
            "Broadcast origin info not found for producer broadcast domain: ",
            p_id->toString(),
            " of ",
            producer->toString());
```

Reply:
I think what we would need to do here is to have `handle(TensorView* tv)` and register new broadcast domains that appear as rfactor domains. That should make sure all broadcast domains in `broadcast_origin_map_`. Once we add some other ways to introduce new broadcast domains, the error check below (at line 103) should trigger. We may want to improve its error message.

Reply:
Ah ok that's a good idea. I will give it a shot. Thanks!

Reply:
I just pushed a change that does what I think you suggested:
- Removes `handle(BroadcastOp*)`
- Removes change to `dispatch(Expr*)`
- Adds `handle(TensorView*)` which looks for rfactor IDs that are broadcast and have definitions, and registers them as origins.

Reply:
I am hitting some errors with this approach
> [E thread_pool.cpp:110] Exception in thread pool task: predicated_parallel_types.none() INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/device_lower/analysis/sync_information.cpp":57, please report a bug with repro script to NVFuser at https://github.co
m/NVIDIA/Fuser/issues. Invalid parallelization of tensor t10. The tensor is parallelized with (threadIdx.x), but it's invalid to use the types as the tensor is also predicated with them., thread pred: (threadIdx.x)   
```
I'll look into this soon.

Reply:
OK I think i get it. The problem is that `BroadcastOp` outputs don't have rfactor domains. So we'll just need to keep the `BroadcastOp` handling method.